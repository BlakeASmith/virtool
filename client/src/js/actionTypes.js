/**
 * Created by igboyes on 03/05/17.
 */

const createRequestActionType = (root) => {
    return {
        REQUESTED: `${root}_REQUESTED`,
        SUCCEEDED: `${root}_SUCCEEDED`,
        FAILED: `${root}_FAILED`
    };
};

// Virus actionTypes
export const WS_UPDATE_VIRUS = "WS_UPDATE_VIRUS";
export const WS_REMOVE_VIRUS = "WS_REMOVE_VIRUS";

export const FIND_VIRUSES = createRequestActionType("FIND_VIRUSES");

export const GET_VIRUS_REQUESTED = "GET_VIRUS_REQUESTED";
export const GET_VIRUS_SUCCEEDED = "GET_VIRUS_SUCCEEDED";
export const GET_VIRUS_FAILED = "GET_VIRUS_FAILED";

export const CREATE_VIRUS_SET_NAME = "CREATE_VIRUS_SET_NAME";
export const CREATE_VIRUS_SET_ABBREVIATION = "CREATE_VIRUS_SET_ABBREVIATION";
export const CREATE_VIRUS_CLEAR = "CREATE_VIRUS_CLEAR";
export const CREATE_VIRUS_REQUESTED = "CREATE_VIRUS_REQUESTED";
export const CREATE_VIRUS_SUCCEEDED = "CREATE_VIRUS_SUCCEEDED";
export const CREATE_VIRUS_FAILED = "CREATE_VIRUS_FAILED";

export const EDIT_VIRUS_REQUESTED = "EDIT_VIRUS_REQUESTED";
export const EDIT_VIRUS_SUCCEEDED = "EDIT_VIRUS_SUCCEEDED";
export const EDIT_VIRUS_FAILED = "EDIT_VIRUS_FAILED";

export const REMOVE_VIRUS_REQUESTED = "REMOVE_VIRUS_REQUESTED";
export const REMOVE_VIRUS_SUCCEEDED = "REMOVE_VIRUS_SUCCEEDED";
export const REMOVE_VIRUS_FAILED = "REMOVE_VIRUS_FAILED";

// History actionTypes
export const FIND_HISTORY_REQUESTED = "FIND_HISTORY_REQUESTED";
export const FIND_HISTORY_SUCCEEDED = "FIND_HISTORY_SUCCEEDED";
export const FIND_HISTORY_FAILED = "FIND_HISTORY";

export const REMOVE_CHANGE_REQUESTED = "REMOVE_CHANGE_REQUESTED";
export const REMOVE_CHANGE_SUCCEEDED = "REMOVE_CHANGE_SUCCEEDED";
export const REMOVE_CHANGE_FAILED = "REMOVE_CHANGE_FAILED";

// Users actionTypes
export const LOAD_USERS = "LOAD_USERS";
export const CREATE_USER = "CREATE_USER";

// Account actionTypes
export const GET_ACCOUNT_REQUESTED = "GET_ACCOUNT_REQUESTED";
export const GET_ACCOUNT_SUCCEEDED = "GET_ACCOUNT_SUCCEEDED";
export const GET_ACCOUNT_FAILED = "GET_ACCOUNT_FAILED";
export const LOGOUT_REQUESTED = "LOGOUT_REQUESTED";
export const LOGOUT_SUCCEEDED = "LOGOUT_SUCCEEDED";

// Administrative setting actionTypes
export const GET_SETTINGS = createRequestActionType("GET_SETTINGS");
export const UPDATE_SETTINGS = createRequestActionType("UPDATE_SETTINGS");

export const SET_SOURCE_TYPE_VALUE = "SET_SOURCE_TYPE_VALUE";
export const SET_CONTROL_READAHEAD_TERM = "SET_CONTROL_READAHEAD_TERM";
export const GET_CONTROL_READAHEAD_REQUESTED = "GET_CONTROL_READAHEAD_REQUESTED";
export const GET_CONTROL_READAHEAD_SUCCEEDED = "GET_CONTROL_READAHEAD_SUCCEEDED";
export const GET_CONTROL_READAHEAD_FAILED = "GET_CONTROL_READAHEAD_FAILED";

export const LIST_USERS = createRequestActionType("LIST_USERS");
export const SELECT_USER = createRequestActionType("SELECT_USER");
export const CHANGE_SET_PASSWORD = "CHANGE_SET_PASSWORD";
export const CLEAR_SET_PASSWORD = "CLEAR_SET_PASSWORD";
export const SET_FORCE_RESET = createRequestActionType("SET_FORCE_RESET");
