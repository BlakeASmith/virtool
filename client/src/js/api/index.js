/**
 *
 *
 * @copyright 2017 Government of Canada
 * @license MIT
 * @author igboyes
 *
 */

export * from "./viruses";
