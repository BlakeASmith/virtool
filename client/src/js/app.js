/**
 * Created by igboyes on 03/03/15.
 */

var React = require('react');
var ReactDOM = require('react-dom');
var Start = require('./components/Start.jsx');

document.write(require('../css/bootstrap.css'));
document.write(require('../css/font.css'));
document.write(require('../css/perfect-scrollbar.css'));
document.write(require('../css/typeahead.css'));
document.write(require('../css/graphics.css'));
document.write(require('../css/style.css'));

ReactDOM.render(React.createElement(Start), document.getElementById('app-container'));