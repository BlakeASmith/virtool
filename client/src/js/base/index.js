import { Alert } from "./Alert";
import { AlgorithmSelect } from "./AlgorithmSelect";
import { AutoProgressBar, ProgressBar } from "./ProgressBar";
import { bsStyles } from "./utils";
import { Button } from "./Button";
import { Checkbox } from "./Checkbox";
import { Flex, FlexItem } from "./Flex";
import { Help } from "./Help";
import { Icon } from "./Icon";
import { Identicon } from "./Identicon";
import { IDRow } from "./IDRow";
import { Input } from "./Input";
import { InputError } from "./InputError";
import { InputSave } from "./InputSave";
import { ListGroupItem } from "./ListGroupItem";
import { LoadingPlaceholder } from "./LoadingPlaceholder";
import { NoneFound } from "./NoneFound";
import { Pagination } from "./Pagination";
import { Radio } from "./Radio";
import { RelativeTime } from "./RelativeTime";
import { RemoveModal } from "./RemoveModal";
import { Tooltip } from "./Tooltip";
import { ViewHeader } from "./ViewHeader";

export {
    Alert,
    AlgorithmSelect,
    AutoProgressBar,
    bsStyles,
    Button,
    Checkbox,
    Flex,
    FlexItem,
    Help,
    Icon,
    Identicon,
    IDRow,
    Input,
    InputError,
    InputSave,
    ListGroupItem,
    LoadingPlaceholder,
    NoneFound,
    Pagination,
    ProgressBar,
    Radio,
    RelativeTime,
    RemoveModal,
    Tooltip,
    ViewHeader
};
