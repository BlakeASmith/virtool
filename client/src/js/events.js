/**
 * Copyright 2015, Government of Canada.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license.
 *
 * @providesModule Events *
 */

import { transform, remove, each } from "lodash";

/**
 * A simple event managing object that can be bound into other objects to make them capable of managing callback and
 * emitting events.
 *
 * @param eventTypes {array} - An array of event names that can be bound and emitted by the Events object.
 * @param parent {object} - A parent object to forward the on, off, and emit functions to.
 * @class
 */
class Events {

    constructor (eventTypes, parent) {

        // Add a property to this.bound for each event type. The key is the eventType and the value is a list that will
        // all bound callbacks for that eventType.
        this.bound = transform(eventTypes, (result, eventType) => result[eventType] = [], {});

        // Bind the following function to a passed parent object to make it directly able to emit events and bind
        // callbacks. Bind the methods to the Events object.
        if (parent) {
            parent.on = this.on.bind(this);
            parent.off = this.off.bind(this);
            parent.emit = this.emit.bind(this);
        }
    }

    /**
     * Bind a callback to an event.
     *
     * @param event {string} - The event to bind the callback to.
     * @param callback {function} - The function to call when the event is emitted.
     * @func
     */
    on (event, callback) {
        if (this.bound.hasOwnProperty(event)) {
            this.bound[event].push(callback);
        } else {
            throw new Error(`Attempted to bind callback to non-existent event "${event}"`);
        }
    }

    /**
     * Unbind a callback from an event.
     *
     * @param event {string} - The event the callback should be unbound from.
     * @param callback {function} - The callback that should be unbound.
     * @func
     */
    off (event, callback) {
        remove(this.bound[event], (candidateCallback) => callback === candidateCallback);
    }

    /**
     * Emits and event. This entails calling every callback function that is bound to the event. Optional data can be
     * passed as an argument to the callback.
     *
     * @param events {array} the name of the event to emit.
     * @param data {object} a data object that can be passed to the bound callbacks when they are called.
     * @func
     */
    emit (events, data) {
        const eventsArray = events instanceof Array ? events: [events];
        eventsArray.forEach((event) => each(this.bound[event], (callback) => callback(data)));
    }
}

export default Events;
