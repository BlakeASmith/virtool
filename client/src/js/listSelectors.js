import { createSelector } from "reselect";
import { map } from "lodash-es";

const getUsers = state => state.users.list;
const getFiles = state => state.files.documents;
const getSubtractions = state => state.subtraction.documents;

const mapIds = (list) => (map(list, entry => entry.id));

export const usersSelector = createSelector(
    [ getUsers ],
    (list) => {

        if (list === null) {
            return { documents: null, page: 0, page_count: 0 };
        }

        return {
            documents: mapIds(list.documents),
            page: list.page,
            page_count: list.page_count
        };

    }
);

export const filesSelector = createSelector([ getFiles ], mapIds);
export const subtractionsSelector = createSelector([ getSubtractions ], mapIds);
