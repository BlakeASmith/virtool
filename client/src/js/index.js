import React from "react";
import ReactDOM from "react-dom";
import Raven from "raven-js";

import App from "./App";
import WSConnection from "./websocket";
import createHistory from "history/createBrowserHistory";
import { getAccount } from "./account/actions";
import { getSettings } from "./settings/actions";
import { assign } from "lodash";
import { createStore, combineReducers, applyMiddleware } from "redux";
import createSagaMiddleware from "redux-saga";
import { routerReducer, routerMiddleware } from "react-router-redux";

import { SET_APP_PENDING, UNSET_APP_PENDING } from "./actionTypes";
import jobsReducer from "./jobs/reducers";
import samplesReducer from "./samples/reducers";
import indexesReducer from "./indexes/reducers";
import virusesReducer from "./viruses/reducers";
import subtractionReducer from "./subtraction/reducers";
import filesReducer from "./files/reducer";
import accountReducer from "./account/reducers";
import settingsReducer from "./settings/reducers";
import usersReducer from "./users/reducers";
import groupsReducer from "./groups/reducers";
import updatesReducer from "./updates/reducers";
import rootSaga from "./sagas";

export * from "../style/style.less";

Raven.config("https://9a2f8d1a3f7a431e873207a70ef3d44d@sentry.io/220532").install();

window.Raven = Raven;

const sagaMiddleware = createSagaMiddleware();

const appInitialState = {
    pending: false
};

const appReducer = (state = appInitialState, action) => {

    switch (action.type) {

        case SET_APP_PENDING:
            return assign({}, state, {pending: true});

        case UNSET_APP_PENDING:
            return assign({}, state, {pending: false});
    }

    return state;
};

const history = createHistory();

const store = createStore(
    combineReducers({
        app: appReducer,
        jobs: jobsReducer,
        samples: samplesReducer,
        viruses: virusesReducer,
        indexes: indexesReducer,
        subtraction: subtractionReducer,
        files: filesReducer,
        settings: settingsReducer,
        users: usersReducer,
        groups: groupsReducer,
        account: accountReducer,
        updates: updatesReducer,
        router: routerReducer
    }),
    applyMiddleware(sagaMiddleware, routerMiddleware(history)),
);

sagaMiddleware.run(rootSaga);

window.store = store;

window.ws = new WSConnection(store.dispatch);
window.ws.establishConnection();

window.store.dispatch(getAccount());
window.store.dispatch(getSettings());

window.addEventListener("beforeunload", () => {
    if (!window.store.getState().files.uploadsComplete) {
        return "hello world";
    }
});

ReactDOM.render(
    <App store={store} history={history} />,
    document.getElementById("app-container")
);
