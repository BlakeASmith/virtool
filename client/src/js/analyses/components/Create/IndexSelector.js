import { map, some } from "lodash-es";
import React, { useCallback } from "react";
import { Label, ListGroup } from "react-bootstrap";
import styled from "styled-components";
import { Checkbox, ListGroupItem, NoneFound } from "../../../base";

const StyledIndexSelectItem = styled(ListGroupItem)`
    align-items: center;
    display: flex;
    justify-content: space-between;

    .label {
        margin-left: 3px;
    }

    span {
        align-items: center;
        display: flex;
    }

    strong {
        margin-left: 8px;
    }
`;

export const IndexSelectItem = ({ id, reference, isSelected, version, onSelect }) => {
    const handleClick = useCallback(() => onSelect({ id, refId: reference.id }), []);
    return (
        <StyledIndexSelectItem onClick={handleClick}>
            <span>
                <Checkbox checked={isSelected} />
                <strong>{reference.name}</strong>
            </span>
            <span>
                Index Version <Label>{version}</Label>
            </span>
        </StyledIndexSelectItem>
    );
};

const StyledIndexSelectList = styled(ListGroup)`
    border: 1px solid ${props => (props.error.length ? "#d44b40" : "transparent")};
    max-height: 165px;
    overflow-y: auto;
    margin-bottom: 3px;
`;

const IndexSelectList = ({ error, indexes, selected, onSelect }) => {
    const indexComponents = map(indexes, index => {
        const isSelected = some(selected, { id: index.id });
        return <IndexSelectItem key={index.id} {...index} isSelected={isSelected} onSelect={onSelect} />;
    });

    return <StyledIndexSelectList error={error}>{indexComponents}</StyledIndexSelectList>;
};

export const IndexSelector = ({ indexes, onSelect, selected, error }) => {
    return (
        <div style={{ marginBottom: "16px" }}>
            <label className="control-label">References</label>
            {indexes.length ? (
                <IndexSelectList error={error} indexes={indexes} selected={selected} onSelect={onSelect} />
            ) : (
                <NoneFound noun="source references" />
            )}
        </div>
    );
};
