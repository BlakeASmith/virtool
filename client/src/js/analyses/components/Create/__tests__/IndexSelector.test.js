import { IndexSelector } from "../IndexSelector.js";

describe("<IndexSelector />", () => {
    let props;

    beforeEach(() => {
        props = {
            indexes: [{ id: "foo" }, { id: "bar" }, { id: "baz" }],
            onSelect: jest.fn(),
            selected: [{ id: "foo" }, { id: "bar" }],
            error: ""
        };
    });

    it("should render", () => {
        const wrapper = shallow(<IndexSelector {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [indexes.length = 0]", () => {
        props.indexes = [];
        const wrapper = shallow(<IndexSelector {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render with error", () => {
        props.error = "Error";
        const wrapper = shallow(<IndexSelector {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
