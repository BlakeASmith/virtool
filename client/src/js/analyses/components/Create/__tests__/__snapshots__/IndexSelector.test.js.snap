// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexSelector /> should render 1`] = `
<div
  style={
    Object {
      "marginBottom": "16px",
    }
  }
>
  <label
    className="control-label"
  >
    References
  </label>
  <IndexSelectorList
    error={
      Array [
        "foo",
      ]
    }
    indexes={
      Array [
        Object {
          "id": "foo",
        },
        Object {
          "id": "bar",
        },
      ]
    }
    onSelect="foo"
    selected={
      Array [
        Object {
          "id": "foo",
        },
        Object {
          "id": "bar",
        },
      ]
    }
  />
</div>
`;
